File: backend/app/services/auth_service.py
from datetime import UTC, datetime, timedelta
from uuid import UUID

from sqlalchemy.ext.asyncio import AsyncSession

from app.core import errors, security
from app.core.config import get_settings
from app.models import User
from app.repositories import auth_repository

settings = get_settings()

ACCESS_COOKIE = "access_token"
REFRESH_COOKIE = "refresh_token"


def _now() -> datetime:
    return datetime.now(UTC)


def build_cookie_params(token: str, max_age: int) -> dict:
    return {
        "value": token,
        "max_age": max_age,
        "httponly": True,
        "samesite": "lax",
        "secure": False,
        "path": "/",
    }


async def authenticate(db: AsyncSession, username: str, password: str) -> User:
    user = await auth_repository.get_user_by_username(db, username)
    if user is None or not security.verify_password(password, user.password_hash):
        raise errors.E_UNAUTHORIZED
    return user


async def issue_refresh_token(db: AsyncSession, user_id: UUID) -> str:
    raw = security.generate_refresh_token()
    expires = _now() + timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS)
    await auth_repository.create_refresh_token(
        db, user_id, security.hash_refresh_token(raw), expires
    )
    return raw


async def rotate_refresh_token(db: AsyncSession, old_raw: str, user_id: UUID) -> str:
    token_hash = security.hash_refresh_token(old_raw)
    stored = await auth_repository.get_refresh_token(db, token_hash)
    if stored is None or stored.revoked or stored.expires_at < _now() or stored.user_id != user_id:
        raise errors.E_REFRESH_EXPIRED
    stored.revoked = True
    raw = security.generate_refresh_token()
    expires = _now() + timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS)
    await auth_repository.create_refresh_token(
        db, user_id, security.hash_refresh_token(raw), expires
    )
    return raw


async def logout(db: AsyncSession, raw_token: str) -> None:
    if raw_token:
        await auth_repository.revoke_refresh_token(db, security.hash_refresh_token(raw_token))

File: backend/internal/service/auth_service.go
package service

import (
	"context"
	"time"

	"angerlog/internal/config"
	"angerlog/internal/models"
	"angerlog/internal/repository"
	"angerlog/internal/security"

	"github.com/google/uuid"
)

const (
	AccessCookie  = "access_token"
	RefreshCookie = "refresh_token"
)

type AuthService struct {
	repo *repository.AuthRepository
	cfg  *config.Config
}

func NewAuthService(repo *repository.AuthRepository, cfg *config.Config) *AuthService {
	return &AuthService{
		repo: repo,
		cfg:  cfg,
	}
}

func (s *AuthService) Authenticate(ctx context.Context, username, password string) (*models.User, *models.AppError) {
	if username == "" || password == "" {
		return nil, models.ErrParam
	}

	user, err := s.repo.GetUserByUsername(ctx, username)
	if err != nil {
		return nil, models.ErrInternal
	}
	if user == nil || !security.VerifyPassword(password, user.PasswordHash) {
		return nil, models.ErrUnauthorized
	}

	return user, nil
}

func (s *AuthService) IssueRefreshToken(ctx context.Context, userID uuid.UUID) (string, *models.AppError) {
	raw := security.GenerateRefreshToken()
	tokenHash := security.HashRefreshToken(raw)
	expiresAt := time.Now().UTC().Add(time.Duration(s.cfg.RefreshTokenExpireDays) * 24 * time.Hour)

	_, err := s.repo.CreateRefreshToken(ctx, userID, tokenHash, expiresAt)
	if err != nil {
		return "", models.ErrInternal
	}

	return raw, nil
}

func (s *AuthService) RotateRefreshToken(ctx context.Context, oldRaw string, userID uuid.UUID) (string, *models.AppError) {
	tokenHash := security.HashRefreshToken(oldRaw)
	stored, err := s.repo.GetRefreshToken(ctx, tokenHash)
	if err != nil {
		return "", models.ErrInternal
	}
	if stored == nil || stored.Revoked || stored.ExpiresAt.Before(time.Now().UTC()) || stored.UserID != userID {
		return "", models.ErrRefreshExpired
	}

	if err := s.repo.RevokeRefreshToken(ctx, tokenHash); err != nil {
		return "", models.ErrInternal
	}

	return s.IssueRefreshToken(ctx, userID)
}

func (s *AuthService) Logout(ctx context.Context, rawToken string) *models.AppError {
	if rawToken != "" {
		tokenHash := security.HashRefreshToken(rawToken)
		_ = s.repo.RevokeRefreshToken(ctx, tokenHash)
	}
	return nil
}

func (s *AuthService) UpdateBottleStyle(ctx context.Context, userID uuid.UUID, style string) (*models.User, *models.AppError) {
	if !models.IsValidBottleStyle(style) {
		return nil, models.ErrParam
	}

	updated, err := s.repo.UpdateBottleStyle(ctx, userID, style)
	if err != nil {
		return nil, models.ErrInternal
	}
	if updated == nil {
		return nil, models.ErrUnauthorized
	}

	return updated, nil
}

func (s *AuthService) GetUserByID(ctx context.Context, id uuid.UUID) (*models.User, *models.AppError) {
	user, err := s.repo.GetUserByID(ctx, id)
	if err != nil {
		return nil, models.ErrInternal
	}
	if user == nil {
		return nil, models.ErrNotFound
	}
	return user, nil
}

func (s *AuthService) GetRefreshToken(ctx context.Context, rawToken string) (*models.RefreshToken, *models.AppError) {
	tokenHash := security.HashRefreshToken(rawToken)
	stored, err := s.repo.GetRefreshToken(ctx, tokenHash)
	if err != nil {
		return nil, models.ErrInternal
	}
	if stored == nil {
		return nil, models.ErrRefreshExpired
	}
	return stored, nil
}

